examples/auth.rs

```diff
@@ -1,45 +1,279 @@
+        class AuthorizationException { * }
+        class AuthorizationExceptionDenied { * }
+    }
+
+    // XX: can be removed when we automatically look through extends/implements
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationException {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationExceptionUnauthenticated {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationExceptionInvalidSecurityToken {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationExceptionInvalidSignature {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthorizationException {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthorizationExceptionDenied {}
+
+    pub use auth::*;
+}
+
+use java_auth::{
+    AuthenticatedExt, AuthenticationExceptionUnauthenticatedExt, AuthorizationExceptionDeniedExt,
+    HttpAuthExt,
 };
 
-duchess::java_package! {
-    package me.ferris;
+pub struct HttpAuth(Global<java_auth::HttpAuth>);
 
-    class AuthenticateResult { * }
-    class HttpAuth { * }
-    class HttpRequest { * }
+#[derive(Debug)]
+pub struct HttpRequest {
+    pub verb: String,
+    pub path: String,
+    pub body_hash: Vec<u8>,
+    pub params: HashMap<String, Vec<String>>,
+    pub headers: HashMap<String, Vec<String>>,
 }
 
-use me::ferris::*;
+pub struct Authenticated {
+    pub account_id: String,
+    pub user: String,
+    state: Global<java_auth::Authenticated>,
+}
 
-fn perform_auth_request() -> duchess::GlobalResult<String> {
-    Jvm::with(|jvm| {
-        let params = HashMap::new().execute(jvm)?;
-        let values = ArrayList::new().execute(jvm)?;
-        values.add("first-value").execute(jvm)?;
-        values.add("second-value").execute(jvm)?;
-        params.put("first-param", &values).execute(jvm)?;
+#[derive(Debug)]
+pub enum AuthenticateError {
+    Unathenticated(String),
+    InvalidSecurityToken,
+    InvalidSignature,
+    InternalError(String),
+}
 
-        let http_request =
-            HttpRequest::new("POST", "/", [1i8, 2, 3].as_slice(), &params).execute(jvm)?;
+#[derive(Debug)]
+pub struct AuthorizeRequest {
+    pub resource: String,
+    pub action: String,
+    pub context: HashMap<String, String>,
+}
 
-        Ok(http_request.to_string().assert_not_null().into_rust(jvm)?)
-    })
+#[derive(Debug)]
+pub enum AuthorizeError {
+    Denied(String),
+    InternalError(String),
 }
 
-fn main() -> duchess::GlobalResult<()> {
-    let s = perform_auth_request()?;
-    println!("{s}");
-    Ok(())
+impl HttpAuth {
+    pub fn new() -> duchess::GlobalResult<Self> {
+        let auth = Jvm::with(|jvm| {
+            let auth = java_auth::HttpAuth::new().execute(jvm)?;
+            Ok(jvm.global(&*auth))
+        })?;
+        Ok(Self(auth))
+    }
+
+    pub fn authenticate(&self, request: &HttpRequest) -> Result<Authenticated, AuthenticateError> {
+        Jvm::with(|jvm| {
+            let request = request.into_java(jvm)?;
```

Comment:
It'd be cool to autogenerate this too, I think, and maybe base it on serialization? Have to think about it.